src/basic/architecture.c

```diff
@@ -150,6 +151,7 @@ static const char *const architecture_table[_ARCHITECTURE_MAX] = {
         [ARCHITECTURE_CRIS]        = "cris",
         [ARCHITECTURE_X86_64]      = "x86-64",
         [ARCHITECTURE_X86]         = "x86",
+        [ARCHITECTURE_X32]         = "x32",
```

Comment:
no, that's not an architecture. It's a calling convention. There are no CPUs that implement this "architecture". The CPUs are just regular x86 cpus.

Also, it's really dead afaik. Let's not add this now.

Reply:
these are not architectures in the cpu sense, they are architectures in the binary abi/package sense, and x32 is a different packaging architecture that is separate and independent from amd64 and i386, and has its own separate builds. So just like you have an amd64/i386 rootfs or usr partition, you also have an x32 partition that ships an installation made of x32 deb packages

Reply:
And it's not dead as a Debian architecture, it is in the same state as other non-release ports like riscv or loongarch or sparc64, from the Debian point of view they are all at the same level of support

Reply:
https://packages.debian.org/sid/x32/systemd/filelist

Reply:
we use the partition types to automatically find the partition to boot into based on efi arch or cpu or kernel arch we find. Nothing of that kind applies to x32. you cannot auto-select that. 

people should just use mark their x32 partitions as x86-64. It makes sense to then mark individual binaries with x32, but the file system as a whole doesn't really make any sense.

There are numerous abis around, including a myriad for arm. we never exposed that either.

Reply:
why not just map x32 to x86_64 on debian?

Reply:
You cannot mark an x32 partition as amd64, because then it would be picked up on amd64 which is wrong. The entire point of x32 is to have a different user space abi than amd64, so they have to be separate, they are not interchangeable

Reply:
And yes the separate selection logic needs some tweaks too - for example using efi arch to select user space can lead to wrong results too, ie there are cursed machines with ia32 firmware and amd64 CPUs that will boot am amd64 kernel and an x32 userspace, because why not

Reply:
> You cannot mark an x32 partition as amd64, because then it would be picked up on amd64 which is wrong. The entire point of x32 is to have a different user space abi than amd64, so they have to be separate, they are not interchangeable

I am not following.

There is no x32 kernel, there's no x32 uefi or anything. you can use the same kernel and boot either 64bit or x32 userspace with it, it's entirely fine. Hence, any auto-discovery of root partition is entirely happy with *either* of these, hence there's no point in having distinct partition types for them.

whether you use x86-64 abi or x32 abi is mostly a choice like "ubuntu" or "fedora". We don't have distinct partition types for that either.

Reply:
These GUIDs match the ABI they contain, not the CPU. Otherwise we would need to always mark an i386 partition as amd64, unless it is actually running on an i386 CPU, as you can take an amd64 CPU, boot an amd64 kernel, and boot an i386 partition. The ABI of the binaries contained in a x32 partition is x32, not amd64 nor i386.

Reply:
Nah, it works the other way round: if systematically the binaries in some partition cannot be run on any other arch then they need their own per-arch id.

hence: because there are plenty binaries that work on x86-64 but do not work on i386 they need a distinct arch type. but there are no real binaries that work on x86-64 but do not work on x32 systems and vice versa, hence they do not need and should not have a distinct identifier.

Reply:
"cannot be run" is a property of an individual system in an individual state, and I cannot possibly know what that will be when building an image beforehand. You can make the same argument for i386: I can run any of that on amd64, hence we can just alias it? Of course not. The only thing I know when I am building an image is, what is the architecture of the content. I do not know on which kernel or which cpu it will be run on.